### PR TITLE
Enable int4mm for both half and bfloat16

### PR DESCRIPTION
**Pull request by malfet**
By making performant kernels a template specialization. This is a prep change for enabling ARM+float16 fast int4 kernel
TODO:
 - Add float32 and some testing

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by mikekgfb — APPROVED** [review]
LGTM

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/native/cpu/int4mm_kernel.cpp [hunk 1]
```diff
@@ -589,6 +589,20 @@ void int4pack_mm_kernel(
...
+    const Tensor& C,
+    const Tensor& A,
+    const Tensor& B,
+    int qGroupSize,
+    const Tensor& qScaleAndZeros,
+    int N, int K) {
+ if (C.scalar_type() == kBFloat16) {
+   int4pack_mm_kernel_<BFloat16>(C, A, B, qGroupSize, qScaleAndZeros, N, K);
+ } else {
+   int4pack_mm_kernel_<Half>(C, A, B, qGroupSize, qScaleAndZeros, N, K);
```
Comment (digantdesai) [inline aten/src/ATen/native/cpu/int4mm_kernel.cpp:602]: This should be safe from overflowing because we do math in f32, right?

**Comment by malfet** [comment]
> @pytorchbot merge

Oops, I wanted to push float32 support into the same PR, but I guess follow-up is coming :)

**Comment by jgong5** [comment]
> Before @malfet wrote this PR, I converted inputs to bf16, and outputs back to the input data type. that worked fine for torch.float, but failed miserably for torch.half as follows below. Is this worthwhile fixing, or we just take this PR and move on?
> 
> xref: #123830
> 
> cc: @jgong5

Oh, ok, it looks something we need to enhance, i.e. supporting the conversion between fp16 and bf16 in the cpp kernel. But I guess we would prefer the kernel to work on fp16 input directly (and internally it could use fp32 as the compute dtype). It's kind of awkward that convert fp16 to bf16 and then use fp32 to compute...

**Comment by malfet** [comment]
> Oh, ok, it looks something we need to enhance, i.e. supporting the conversion between fp16 and bf16 in the cpp kernel. But I guess we would prefer the kernel to work on fp16 input directly (and internally it could use fp32 as the compute dtype). It's kind of awkward that convert fp16 to bf16 and then use fp32 to compute...

I agree, that's the pattern we should avoid in the codebase, but compiler is the compiler and should support any user code thrown at it

**Comment by mikekgfb** [comment]
> > Oh, ok, it looks something we need to enhance, i.e. supporting the conversion between fp16 and bf16 in the cpp kernel. But I guess we would prefer the kernel to work on fp16 input directly (and internally it could use fp32 as the compute dtype). It's kind of awkward that convert fp16 to bf16 and then use fp32 to compute...
> 
> I agree, that's the pattern we should avoid in the codebase, but compiler is the compiler and should support any user code thrown at it

+1.  Totally agree, it atrocious as an idiom, and as a least common denominator it has neither a useful mantissa nor exponent.   
But a compile should not die, although the generated code might not be great (best is probably convert to fp32 and then to the other format), but I don't think users should have an expectation that this would perform well.  (I personally only put together that example because I know @malfet was working on the operator and wanted to get a head start on enabling these execution paths for mobile where float16 rather than bf16 is still the norm.)

**Pull request closed without merging**